### PR TITLE
Add `contrib/upgrade.sh`

### PR DESCRIPTION
**Pull request by noerw**
Adds an upgrade script that automates upgrading installations on Linux from binary releases, so people don't need to reinvent the wheel. Hopefully this leads to less questions about how to upgrade, and consequently less Gitea instances running unmaintained versions in the wild.

Feel free to suggest improvements.
originally posted in https://github.com/go-gitea/gitea/issues/17405#issuecomment-949777985

edit: This is meant as a stopgap for common linux deployments, an updater supporting all the different ways gitea can be deployed probably has ~2 orders of magnitude more complexity, so I didn't even try to accomodate for that.
Even for differing linux environments I assumed the user screens/tests/adapts this script for their needs, but I'll implement your feedback to make it more portable later (though given the tradeoff between portability and readability/extensibility, I side more with the latter option).

### CODE REVIEW

File: contrib/upgrade.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
#!/usr/bin/env bash
```

More portable for linux distro's and systems. [See](https://stackoverflow.com/questions/16365130/what-is-the-difference-between-usr-bin-env-bash-and-usr-bin-bash#16365367)

File: contrib/upgrade.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+set -e
+
+# this is an upgrade script, for gitea deployed on linux as systemd service
+# depends on: curl, xz, sha256sum, gpg
+# usage:      upgrade.sh [version]
+
+# change the variables below for your local setup
+giteaversion=${1:-1.15.10}
```
Comment (Gusted) [inline contrib/upgrade.sh]: Should `-1.15.10` be updated every version?

File: contrib/upgrade.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+giteabin=/usr/local/bin/gitea
+giteahome=/var/lib/gitea
+giteaconf=/etc/gitea/app.ini
+giteauser="git"
```
Comment (Gusted) [inline contrib/upgrade.sh]: I think this is too much of a assumption, They are great defaults - but we should give them a option to override it.

File: contrib/upgrade.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+giteacmd="sudo -u $giteauser $giteabin -c $giteaconf -w $giteahome"
```
Comment (Gusted) [inline contrib/upgrade.sh]: `sudo` is the most popular privilege escape tool, `doas` has been in adoption as well. So we shouldn't assume `sudo` is available. Check for `doas` first, then `sudo`.

File: contrib/upgrade.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+# download new binary
+binname=gitea-${giteaversion}-linux-amd64
+binurl="https://dl.gitea.io/gitea/${giteaversion}/${binname}.xz"
+echo downloading $binurl
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
echo Downloading "$binurl"
```
Just to avoid some bash issues in case.

File: contrib/upgrade.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+cd $giteahome # needed for gitea dump later
```
Comment (Gusted) [inline contrib/upgrade.sh]: Maybe re-order this? Such that the echo is afterwards this `cd`, `cd` could fail because of a non-existent path on the system. Which would be weird when getting a downloading notice before-hand 😅.

File: contrib/upgrade.sh [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+curl -sSfL "$binurl" > ${binname}.xz
```
Comment (Gusted) [inline contrib/upgrade.sh]: I think we should prefer using `-o "${binname}.xz"` instead of using the `>`.  Ditto that for the 2 other curl commands as well.

Comment (Gusted) [reply contrib/upgrade.sh]: Also, we could simplify it to 1 command by using `curl`'s globbing abilities.

File: contrib/upgrade.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+curl -sSfL "${binurl}.sha256" > ${binname}.xz.sha256
+curl -sSfL "${binurl}.asc" > ${binname}.xz.asc
+
+# validate checksum & gpg signature (exit script if error)
+sha256sum -c ${binname}.xz.sha256
+gpg --keyserver keys.openpgp.org --recv 7C9E68152594688862D62AF62D9AE806EC1592E2
```
Comment (Gusted) [inline contrib/upgrade.sh:63]: Kind of a concern. Not sure if we should do a gpg import keys in a shell script. Also, `7C9E68152594688862D62AF62D9AE806EC1592E2`'s keys has a expiration(which is btw not a good thing for signing keys)  so in the case they don't push a non-expiring key to the keyservers and create a new key, this script will fail.

File: contrib/upgrade.sh [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+rm ${binname}.xz.{sha256,asc}
+
+# unpack binary + make executable
+xz -d ${binname}.xz
+chmod +x $binname
+
+# stop gitea, create backup, replace binary, restart gitea
+$giteacmd manager flush-queues
+systemctl stop gitea
+$giteacmd --version
```
Comment (Gusted) [inline contrib/upgrade.sh]: Necessary?

File: contrib/upgrade.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+# validate checksum & gpg signature (exit script if error)
+sha256sum -c ${binname}.xz.sha256
+gpg --keyserver keys.openpgp.org --recv 7C9E68152594688862D62AF62D9AE806EC1592E2
+gpg --verify ${binname}.xz.asc ${binname}.xz
+rm ${binname}.xz.{sha256,asc}
+
+# unpack binary + make executable
+xz -d ${binname}.xz
+chmod +x $binname
```
Comment (Gusted) [inline contrib/upgrade.sh]: We have a gitea user(those who use proper isolation should have a separate user), maybe chown and give them the permissions to the file?

File: contrib/upgrade.sh [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+# stop gitea, create backup, replace binary, restart gitea
+$giteacmd manager flush-queues
+systemctl stop gitea
+$giteacmd --version
+$giteacmd dump
+mv -fb $binname $giteabin
```
Comment (Gusted) [inline contrib/upgrade.sh]: is `f` necessary, I don't like to use `force` by-default in a shell script.

File: contrib/upgrade.sh (hunk 5 above)
Comment (Gusted) [reply contrib/upgrade.sh]: Ditto this for all other variable use-cases.

File: contrib/upgrade.sh [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+gpg --verify ${binname}.xz.asc ${binname}.xz
+rm ${binname}.xz.{sha256,asc}
+
+# unpack binary + make executable
+xz -d ${binname}.xz
+chmod +x $binname
+
+# stop gitea, create backup, replace binary, restart gitea
+$giteacmd manager flush-queues
+systemctl stop gitea
```
Comment (Gusted) [inline contrib/upgrade.sh]: Not sure in which environment we could expect this to run, but it might be necessary to sudo/doas this one as root.

**Review by Gusted — COMMENTED** [review]
I can see how this is helpful for servers to fast upgrade their gitea instance. But the script needs some work to do, in order to avoid some issues.

File: contrib/upgrade.sh (hunk 6 above)
Comment (delvh) [reply contrib/upgrade.sh]: Also, I think
```suggestion
pushd $(pwd);
cd $giteahome # needed for gitea dump later
```
would be nice, especially with a `popd` at the end.

File: contrib/upgrade.sh (hunk 2 above)
Comment (delvh) [reply contrib/upgrade.sh]: An approach with less human overhead would probably be to use something like
```sh
giteaversion=${1:-$(curl -s --connect-timeout 10 'https://storage.gitea.io/gitea-artifacts/gitea/version.json')}
```
I've tested it, it works (in theory), there are, however, three things to keep in mind:
Firstly, a valid URL key needs to be present as otherwise the access will be denied.
Secondly, the resulting json must somehow be parsed using bash.
Lastly, the curl command can most likely be polished.

File: contrib/upgrade.sh (hunk 5 above)
Comment (delvh) [reply contrib/upgrade.sh]: Alternatively,
```suggestion
echo "Downloading $binurl"
```

File: contrib/upgrade.sh [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+$giteacmd --version
+$giteacmd dump
```
Comment (delvh) [inline contrib/upgrade.sh]: Perhaps there should be a branch here that differentiates between if another variable is given, and if that is the case it calls `$giteacmd dump -c <variable>`, and otherwise simply dumps the items.

**Comment by lunny** [comment]
I can accept this in `contrib` but an official updater is better. So that, we need to know the environment, which OS is running the Gitea running and if it's in docker.

File: contrib/upgrade.sh (hunk 8 above)
Comment (noerw) [reply contrib/upgrade.sh:63]: Educate me if I'm wrong; my understanding is that importing a key just makes the key available locally for validation. It doesn't set a trust level on the key or anything. What is your concern?
I suspect that users who don't check this script to validate the key (by crosschecking with the docs - currently there is no better non-human source of truth afaik), wouldn't have checked the signature properly anyway. And if they trust the docs, they can equally trust this script, both live in the same repo.
About expiration: there is nothing I can change here; this is the key binaries are signed with. I'll add a `TODO` comment noting the expiration date (2022-06-24)

File: contrib/upgrade.sh (hunk 10 above)
Comment (noerw) [reply contrib/upgrade.sh]: You mean `chown "$giteauser" "$binname"; chmod 100 "$binname"`?

File: contrib/upgrade.sh (hunk 11 above)
Comment (noerw) [reply contrib/upgrade.sh]: I expect that users express their intent to update gitea by executing this script, I'm not sure why it's meaningful to prompt (for this particular step) again.
We could of course prompt before doing the actual upgrade `are you sure to update from $oldver to $giteaver? you really should check the changelog beforehand: $changelogurl`

Comment (Gusted) [reply contrib/upgrade.sh]: Woops yeah, nvm my message.

File: contrib/upgrade.sh (hunk 10 above)
Comment (Gusted) [reply contrib/upgrade.sh]: Yes.
`100` should be `+x`.

File: contrib/upgrade.sh (hunk 8 above)
Comment (Gusted) [reply contrib/upgrade.sh:63]: So, yes on it's own importing a GPG key isn't a concern. However, certain systems are keen on importing GPG as you still can verify files and keys with. As well certain configurations will have a custom keyserver. I think it's best to mention it.

As for the expiration yeah, @go-gitea/owners will a new key be generated once this one is  expired?

File: contrib/upgrade.sh (hunk 2 above)
Comment (noerw) [reply contrib/upgrade.sh]: good idea, I implemented and added an optional `jq` dependency

File: contrib/upgrade.sh (hunk 3 above)
Comment (noerw) [reply contrib/upgrade.sh]: I made it so you can override a la `giteauser=foo contrib/upgrade.sh`, I think that should do?

File: contrib/upgrade.sh [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+#   upgrade.sh 1.15.10
+#   giteahome=/opt/gitea giteaconf=$giteahome/app.ini upgrade.sh
+
+# apply variables from environment
+: ${giteabin:=/usr/local/bin/gitea}
+: ${giteahome:=/var/lib/gitea}
+: ${giteaconf:=/etc/gitea/app.ini}
+: ${giteauser:=git}
+: ${sudocmd:=sudo}
+: ${arch:=linux-amd64}
```
Comment (delvh) [inline contrib/upgrade.sh]: Where is the `:` coming from?

File: contrib/upgrade.sh [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  for exe in $@; do
+    which $exe &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
+  done
+}
+require curl xz sha256sum gpg
+
+# select version to install
+if [[ -z "$1" ]]; then
+  require jq
+	giteaversion=`curl -sL https://dl.gitea.io/gitea/version.json | jq -r .latest.version`
```
Comment (delvh) [inline contrib/upgrade.sh]:
```suggestion
	giteaversion=`curl -sL --connect-timeout $connecttimeout 'https://dl.gitea.io/gitea/version.json' | jq -r .latest.version`
```

File: contrib/upgrade.sh [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+else
+	giteaversion="${1}"
+fi
+
+# confirm update
+current=`giteacmd --version | cut -d' ' -f3`
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
current=$(giteacmd --version | cut -d' ' -f3)
```
Let's not mix styles. I think we should avoid ` as it's legacy.

File: contrib/upgrade.sh [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+echo "make sure to read the changelog first: https://github.com/go-gitea/gitea/blob/main/CHANGELOG.md"
+echo "are you ready to update Gitea from ${current} to ${giteaversion}? (y/N)"
+read confirm
+[[ "$confirm" == "y" ]] || exit 1
+
+pushd `pwd`
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
pushd $(pwd)
```

File: contrib/upgrade.sh (hunk 16 above)
Comment (delvh) [reply contrib/upgrade.sh]: Also, I just noticed that it would be beneficial to add a check here whether `"$current" == "$giteaversion"` to prevent misuse/ no-ops.
In this specific case, it doesn't even matter whether `"$current" > "$giteaversion"`, because this script in the current implementation can be also used to downgrade gitea.
(Not even an additional warning is needed as the user will already be asked for confirmation below anyway)

File: contrib/upgrade.sh (hunk 14 above)
Comment (delvh) [inline contrib/upgrade.sh]:
```suggestion
: ${arch:=linux-amd64}
${connecttimeout:=10}
```

File: contrib/upgrade.sh [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+cd $giteahome # needed for gitea dump later
+
+# download new binary
+binname=gitea-${giteaversion}-${arch}
+binurl="https://dl.gitea.io/gitea/${giteaversion}/${binname}.xz"
+echo "Downloading $binurl..."
+curl -sSfLO "$binurl{,.sha256,.asc}"
```
Comment (delvh) [inline contrib/upgrade.sh]:
```suggestion
curl --connect-timeout $connecttimeout -sSfLO "$binurl{,.sha256,.asc}"
```

File: contrib/upgrade.sh (hunk 14 above)
Comment (noerw) [reply contrib/upgrade.sh]: this is a bash builtin noop. if omitted, bash will try to execute the assigned value :grimacing:
`giteabin=${giteabin:-/usr/local/bin/gitea}` is probably more readable.

File: contrib/upgrade.sh (hunk 4 above)
Comment (noerw) [reply contrib/upgrade.sh]: :+1: can now be overriden using the `sudocmd` env var.

File: contrib/upgrade.sh (hunk 8 above)
Comment (noerw) [reply contrib/upgrade.sh:63]: Ah ok, I get it now; people are afraid that having a key imported can make `gpg --verify` succeed on unrelated files.
We could drop the key after use again, not sure about how to proceed here :thinking:

Comment (Gusted) [reply contrib/upgrade.sh:63]: > Ah ok, I get it now; people are afraid that having a key imported can make `gpg --verify` succeed on unrelated files. We could drop the key after use again, not sure about how to proceed here thinking

I think it just should be mentioned and then let it be as-is.

File: contrib/upgrade.sh [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+#   giteahome=/opt/gitea giteaconf=$giteahome/app.ini upgrade.sh
+
+# apply variables from environment
+: ${giteabin:=/usr/local/bin/gitea}
+: ${giteahome:=/var/lib/gitea}
+: ${giteaconf:=/etc/gitea/app.ini}
+: ${giteauser:=git}
+: ${sudocmd:=sudo}
+: ${arch:=linux-amd64}
+: ${backupopts:=--skip-lfs-data --skip-repository --skip-attachment-data --skip-log}
```
Comment (delvh) [inline contrib/upgrade.sh]: What exactly is the reason for excluding these things from the backup by default?

I mean, why would that be wanted? To save some space and potentially lose all that data?
Or will those data not be changed by the upgrade and hence can be ignored?

Comment (noerw) [reply contrib/upgrade.sh]: The idea was to speed it up to save downtime & storage space. But yeah it's not a good default, I'll revert this.

File: contrib/upgrade.sh [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+  giteaversion="$1"
+fi
+
+# confirm update
+current=$(giteacmd --version | cut --delimiter=' ' --fields=3)
+[[ "$current" == "$giteaversion" ]] && echo "$current is already installed, stopping." && exit 1
+echo "Make sure to read the changelog first: https://github.com/go-gitea/gitea/blob/main/CHANGELOG.md"
+echo "Are you ready to update Gitea from ${current} to ${giteaversion}? (y/N)"
+read confirm
+[[ "$confirm" == "y" ]] || exit 1
```
Comment (delvh) [inline contrib/upgrade.sh]:
```suggestion
[[ "$confirm" == "y" ]] || [[ "$confirm" == "Y" ]] || exit 1
```

**Review by delvh — APPROVED** [review]

**Review by delvh — APPROVED** [review]

File: contrib/upgrade.sh [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+function giteacmd {
+  "$sudocmd" --user "$giteauser" "$giteabin" --config "$giteaconf" --work-path "$giteahome" $@
+}
+
+function require {
+  for exe in $@; do
+    which $exe &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
+  done
+}
+require curl xz sha256sum gpg
```
Comment (eeyrjmr) [inline contrib/upgrade.sh]: systemctl should be added to this list as this script depends on a linux distribution using systemd as its service manager.
I have gitea running in multiple locations and none use systemd (gentoo, alpine-metal, alpine containers) but gitea is also managed via the package manager (be it v1.15)

Comment (Gusted) [reply contrib/upgrade.sh]: I think ` on linux as systemd service.` in the description of this script should be sufficient. If systemd is installed systemctl is installed as well. But it could be added as a safeguard, but people using things like init should have their own kind of way to do this, as we don't have a ready to go script for such configurations.

**Review by Gusted — APPROVED** [review]

**Comment by silverwind** [comment]
Please lint the script on https://www.shellcheck.net/, it has a lot of warnings:

````
Line 17:
: ${giteabin:="/usr/local/bin/gitea"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 18:
: ${giteahome:="/var/lib/gitea"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 19:
: ${giteaconf:="/etc/gitea/app.ini"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 20:
: ${giteauser:="git"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 21:
: ${sudocmd:="sudo"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 22:
: ${arch:="linux-amd64"}
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 23:
: ${backupopts:=""} # see `gitea dump --help` for available options
  ^-- SC2223 (info): This default assignment may cause DoS due to globbing. Quote it.
 
Line 26:
  "$sudocmd" --user "$giteauser" "$giteabin" --config "$giteaconf" --work-path "$giteahome" $@
>>                                                                                          ^-- SC2068 (error): Double quote array expansions to avoid re-splitting elements.
 
Line 30:
  for exe in $@; do
             ^-- SC2068 (error): Double quote array expansions to avoid re-splitting elements.
 
Line 31:
    which $exe &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
          ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
    which "$exe" &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
 
Line 49:
read confirm
^-- SC2162 (info): read without -r will mangle backslashes.
 
Line 52:
pushd $(pwd)
      ^-- SC2046 (warning): Quote this to prevent word splitting.
 
Line 62:
sha256sum --check ${binname}.xz.sha256
                  ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
sha256sum --check "${binname}".xz.sha256
 
Line 65:
gpg --verify ${binname}.xz.asc ${binname}.xz
             ^-- SC2086 (info): Double quote to prevent globbing and word splitting.
                               ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
gpg --verify "${binname}".xz.asc "${binname}".xz
 
Line 66:
rm ${binname}.xz.{sha256,asc}
   ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
rm "${binname}".xz.{sha256,asc}
 
Line 69:
xz --decompress ${binname}.xz
                ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
xz --decompress "${binname}".xz
````

File: contrib/upgrade.sh [hunk 22]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+set -e
```
Comment (silverwind) [inline contrib/upgrade.sh]:
```suggestion
set -euo pipefail
```

http://redsymbol.net/articles/unofficial-bash-strict-mode/

In some scripts I omit the `e` option in case a sub-command is expected to fail.

**Review by zeripath — DISMISSED** [review]
as per @silverwind

File: contrib/upgrade.sh (hunk 22 above)
Comment (noerw) [reply contrib/upgrade.sh]: good point, done!

**Review by silverwind — APPROVED** [review]

**Comment by noerw** [comment]
found quoting issues in the script since the shellcheck changes, let me check

File: contrib/upgrade.sh [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+: "${arch:="linux-amd64"}"
+: "${backupopts:=""}" # see `gitea dump --help` for available options
+
+function giteacmd {
+  "$sudocmd" --user "$giteauser" "$giteabin" --config "$giteaconf" --work-path "$giteahome" "$@"
+}
+
+function require {
+  for exe in "$@"; do
+    which "$exe" &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
    command -v "$exe" &>/dev/null || (echo "missing dependency '$exe'"; exit 1)
```

[See](https://lwn.net/Articles/874049/)

TL;DR `command -v` is POSIX-compliant, `which` isn't.

File: contrib/upgrade.sh [hunk 24]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# This is an update script for gitea installed via the binary distribution
+# from dl.gitea.io on linux as systemd service. It performs a backup and updates
+# Gitea in place.
+# NOTE: This adds the GPG Signing Key of the Gitea maintainers to the keyring.
+# Depends on: bash, curl, xz, sha256sum, gpg, which. optionally jq.
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
# Depends on: bash, curl, xz, sha256sum, gpg. optionally jq.
```

File: contrib/upgrade.sh [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+# Examples:
+#   upgrade.sh 1.15.10
+#   giteahome=/opt/gitea giteaconf=$giteahome/app.ini upgrade.sh
+
+# apply variables from environment
+: "${giteabin:="/usr/local/bin/gitea"}"
+: "${giteahome:="/var/lib/gitea"}"
+: "${giteaconf:="/etc/gitea/app.ini"}"
+: "${giteauser:="git"}"
+: "${sudocmd:="sudo"}"
```
Comment (Gusted) [inline contrib/upgrade.sh:21]: I'm not sure on this, I think this can just be hardcoded for `sudo`/`doas`.

File: contrib/upgrade.sh [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+binname="gitea-${giteaversion}-${arch}"
+binurl="https://dl.gitea.io/gitea/${giteaversion}/${binname}.xz"
+echo "Downloading $binurl..."
+curl --connect-timeout 10 --silent --show-error --fail --location -O "$binurl{,.sha256,.asc}"
+
+# validate checksum & gpg signature (exit script if error)
+sha256sum --check "${binname}.xz.sha256"
+# TODO 2022-06-24: this gpg key will expire!
+gpg --keyserver keys.openpgp.org --recv 7C9E68152594688862D62AF62D9AE806EC1592E2
+gpg --verify "${binname}.xz.asc" "${binname}.xz"
```
Comment (eeyrjmr) [inline contrib/upgrade.sh]: gpg --verify "${binname}.xz.asc" "${binname}.xz" || { echo 'Signature does not match' ; exit 1; }

File: contrib/upgrade.sh [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+rm "${binname}".xz.{sha256,asc}
+
+# unpack binary + make executable
+xz --decompress "${binname}.xz"
+chown "$giteauser" "$binname"
+chmod +x "$binname"
+
+# stop gitea, create backup, replace binary, restart gitea
+echo "Stopping gitea at $(date)"
+giteacmd manager flush-queues
```
Comment (eeyrjmr) [inline contrib/upgrade.sh:74]: "$giteacmd" manager flush-queues

File: contrib/upgrade.sh [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+$sudocmd systemctl stop gitea
+echo "Creating backup in $giteahome"
+giteacmd dump $backupopts
```
Comment (eeyrjmr) [inline contrib/upgrade.sh:77]: "$giteacmd" dump "$backupopts"

File: contrib/upgrade.sh [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+
+# select version to install
+if [[ -z "${1:-}" ]]; then
+  require jq
+  giteaversion=$(curl --connect-timeout 10 -sL https://dl.gitea.io/gitea/version.json | jq -r .latest.version)
+else
+  giteaversion="$1"
+fi
+
+# confirm update
```
Comment (eeyrjmr) [inline contrib/upgrade.sh:44]: #confirm update
arch_array=("linux-386" "linux-amd64" "linux-arm" "linux-arm64") 
[[ "${arch_array[*]}" =~ "${arch}" ]] ||  { echo "arch variable is not set to a supported string" ; exit 1; }

File: contrib/upgrade.sh [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+current=$(giteacmd --version | cut --delimiter=' ' --fields=3)
+[[ "$current" == "$giteaversion" ]] && echo "$current is already installed, stopping." && exit 1
+echo "Make sure to read the changelog first: https://github.com/go-gitea/gitea/blob/main/CHANGELOG.md"
+echo "Are you ready to update Gitea from ${current} to ${giteaversion}? (y/N)"
+read -r confirm
+[[ "$confirm" == "y" ]] || [[ "$confirm" == "Y" ]] || exit 1
```
Comment (eeyrjmr) [inline contrib/upgrade.sh:50]: [[ "$confirm" =~ ^[Yy]$ ]] ||  exit 1

File: contrib/upgrade.sh [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+else
+  giteaversion="$1"
+fi
+
+# confirm update
+current=$(giteacmd --version | cut --delimiter=' ' --fields=3)
+[[ "$current" == "$giteaversion" ]] && echo "$current is already installed, stopping." && exit 1
+echo "Make sure to read the changelog first: https://github.com/go-gitea/gitea/blob/main/CHANGELOG.md"
+echo "Are you ready to update Gitea from ${current} to ${giteaversion}? (y/N)"
+read -r confirm
```
Comment (eeyrjmr) [inline contrib/upgrade.sh:49]: read -p confirm

Comment (Gusted) [reply contrib/upgrade.sh:49]: I don't think `read` has such option, at least the manual doesn't mention such option being available.
https://linux.die.net/man/2/read

Comment (wxiaoguang) [reply contrib/upgrade.sh:49]: `-p` is incorrect here, `-p` is used to show a prompt message.

```
$ bash -c "help read"
read: read [-ers] [-u fd] [-t timeout] [-p prompt] [-a array] [-n nchars] [-d delim] [name ...]
    If the -r option is given, this signifies `raw' input, and
    backslash escaping is disabled. 
    If the -p option is supplied, the string PROMPT is output without a trailing newline
    before attempting to read. 
```

File: contrib/upgrade.sh (hunk 30 above)
Comment (Gusted) [reply contrib/upgrade.sh:50]: I think the current code is better, it's more clear.

The suggested code is more compact and, took me a while to figure it was actually regex expression

File: contrib/upgrade.sh [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+
+# download new binary
+binname="gitea-${giteaversion}-${arch}"
+binurl="https://dl.gitea.io/gitea/${giteaversion}/${binname}.xz"
+echo "Downloading $binurl..."
+curl --connect-timeout 10 --silent --show-error --fail --location -O "$binurl{,.sha256,.asc}"
+
+# validate checksum & gpg signature (exit script if error)
+sha256sum --check "${binname}.xz.sha256"
+# TODO 2022-06-24: this gpg key will expire!
```
Comment (Gusted) [inline contrib/upgrade.sh]:
```suggestion
```

File: contrib/upgrade.sh (hunk 29 above)
Comment (Gusted) [reply contrib/upgrade.sh:44]: This would make sense if this was a install script... But this assumes that it already has a official installation going on.

File: contrib/upgrade.sh [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+# download new binary
+binname="gitea-${giteaversion}-${arch}"
+binurl="https://dl.gitea.io/gitea/${giteaversion}/${binname}.xz"
+echo "Downloading $binurl..."
+curl --connect-timeout 10 --silent --show-error --fail --location -O "$binurl{,.sha256,.asc}"
+
+# validate checksum & gpg signature (exit script if error)
+sha256sum --check "${binname}.xz.sha256"
+gpg --keyserver keys.openpgp.org --recv 7C9E68152594688862D62AF62D9AE806EC1592E2
+gpg --verify "${binname}.xz.asc" "${binname}.xz"
```
Comment (6543) [inline contrib/upgrade.sh]:
```suggestion
gpg --verify "${binname}.xz.asc" "${binname}.xz" || { echo 'Signature does not match'; exit 1; }
```

**Review by 6543 — APPROVED** [review]

**Comment by HorlogeSkynet** [comment]
Hi @noerw, just a quick message to send you <https://gist.github.com/HorlogeSkynet/d676b9204869842933169dbe35ed8650> (a personal script I maintain publicly for this purpose).
Bye :wave:

**Comment by 6543** [comment]
:robot: wake up
